src/ray/raylet/runtime_env_agent_client.cc

```diff
@@ -0,0 +1,395 @@
+    if (ec) {
+      fail_callback_(ray::Status::NotFound("on_resolve " + ec.message()));
+      return;
+    }
+
+    // Make the connection on the IP address we get from a lookup
+    stream_.async_connect(
+        results, beast::bind_front_handler(&Session::on_connect, shared_from_this()));
+  }
+
+  void on_connect(beast::error_code ec, tcp::resolver::results_type::endpoint_type) {
+    if (ec) {
+      fail_callback_(ray::Status::NotFound("on_connect " + ec.message()));
+      return;
+    }
+
+    // Send the HTTP request to the remote host
+    http::async_write(
+        stream_, req_, beast::bind_front_handler(&Session::on_write, shared_from_this()));
+  }
+
+  void on_write(beast::error_code ec, std::size_t bytes_transferred) {
+    boost::ignore_unused(bytes_transferred);
+
+    if (ec) {
+      fail_callback_(ray::Status::IOError("on_write " + ec.message()));
+      return;
+    }
+
+    // Receive the HTTP response
+    http::async_read(stream_,
+                     buffer_,
+                     res_,
+                     beast::bind_front_handler(&Session::on_read, shared_from_this()));
+  }
+
+  void on_read(beast::error_code ec, std::size_t bytes_transferred) {
+    boost::ignore_unused(bytes_transferred);
+
+    if (ec) {
+      fail_callback_(ray::Status::IOError("on_read " + ec.message()));
+      return;
+    }
+    if (http::to_status_class(res_.result()) == http::status_class::successful) {
+      succ_callback_(std::move(res_).body());
+    } else {
+      fail_callback_(ray::Status::IOError(" POST result non-ok status code " +
+                                          std::to_string(res_.result_int())));
+    }
+
+    // Gracefully close the socket
+    stream_.socket().shutdown(tcp::socket::shutdown_both, ec);
+    // not_connected happens sometimes so don't bother reporting it.
+    if (ec && ec != beast::errc::not_connected) {
+      RAY_LOG(INFO) << "on_read error after response body reseived: " << ec.message();
+    }
+  }
+};
+
+class HttpRuntimeEnvAgentClient : public RuntimeEnvAgentClient {
+ public:
+  HttpRuntimeEnvAgentClient(instrumented_io_context &io_context,
+                            const std::string &address,
+                            int port,
+                            std::function<std::shared_ptr<boost::asio::deadline_timer>(
+                                std::function<void()>, uint32_t delay_ms)> delay_executor)
+      : io_context_(io_context),
+        address_(address),
+        port_str_(std::to_string(port)),
+        delay_executor_(delay_executor) {}
+  ~HttpRuntimeEnvAgentClient() {}
+
+  template <typename T>
+  using SuccCallback = std::function<void(T)>;
+  using FailCallback = std::function<void(ray::Status)>;
+  template <typename T>
+  using TryInvokeOnce = std::function<void(SuccCallback<T>, FailCallback)>;
+
+  /// @brief Invoke `try_invoke_once`. If it fails with a retriable error, retries after
```

Comment:
Do we infinitely retry? Should we add a limit here?

Reply:
Current behavior is: https://github.com/ray-project/ray/blob/40755d6f50ecf454df4774ff127d2237456450e9/src/ray/raylet/agent_manager.cc#L188

- In ctor: starts the agent and waits for registration from the agent
- If registration does not happen after a `agent_register_timeout_ms`, raylet suicides.
- On the 3 methods call: indefinitely retries until the agent registered.

But we don't have a registration now. Maybe we can do this:
- On ctor: (nothing)
- On the 3 methods call: retries until `agent_register_timeout_ms` passed, after which raylet suicides.

WDYT? Added related code.

Reply:
per offline discussion, added suicide code.

Reply:
Yeah I like the new proposal!